mlir/lib/Analysis/Presburger/Matrix.cpp

```diff
@@ -432,4 +432,111 @@ MPInt IntMatrix::normalizeRow(unsigned row, unsigned cols) {
 
 MPInt IntMatrix::normalizeRow(unsigned row) {
   return normalizeRow(row, getNumColumns());
+}
+
+MPInt IntMatrix::determinant(IntMatrix* inverse) const {
+  unsigned nRows = getNumRows();
+  unsigned nColumns = getNumColumns();
+
+  assert(nRows == nColumns && "determinant can only be calculated for square matrices!");
+
+  FracMatrix m(*this);
+  
+  FracMatrix fracInverseM(nRows, nColumns);
+  FracMatrix* fracInverse = &fracInverseM;
+  Fraction detF = m.determinant(fracInverse);
+
+  MPInt detM = detF.getAsInteger();
+
+  if (detM != 0) {
+    *inverse = IntMatrix(nRows, nColumns);
+    for (unsigned i = 0; i < nRows; i++)
+      for (unsigned j = 0; j < nColumns; j++)
+        inverse->at(i, j) = (fracInverse->at(i, j) * detF).getAsInteger();
+  }
+
+  return detM;
+}
+
+FracMatrix FracMatrix::identity(unsigned dimension) {
+  return Matrix::identity(dimension);
+}
+
+FracMatrix::FracMatrix(IntMatrix m)
+  : FracMatrix(m.getNumRows(), m.getNumColumns()) {
+  for (unsigned i = 0; i < m.getNumRows(); i++)
+    for (unsigned j = 0; j < m.getNumColumns(); j++)
+      this->at(i, j) = Fraction(m.at(i, j), 1);
+}
+
+Fraction FracMatrix::determinant(FracMatrix* inverse) const {
+  unsigned nRows = getNumRows();
+  unsigned nColumns = getNumColumns();
+
+  assert(nRows == nColumns && "determinant can only be calculated for square matrices!");
+
+  FracMatrix m(*this);
+  FracMatrix tempInv(nRows, nColumns);
+  if (inverse) tempInv = FracMatrix::identity(nRows);
+
+  Fraction a, b;
+  // For each row in the matrix,
```

Comment:
please give an overview at the top of the loop. We can see that the loop is over the rows of the matrix. instead of that, add high-level information like 

// Make the matrix into lower triangular form using gaussian elimination with row operations.
// If inverse is required, we apply more operations to turn the matrix into diagonal form and apply
// the same operations to the inverse matrix that was initially identity. Either way, the product of the
// diagonal elements is then the determinant.

Some duplication with the comments inside the loop is ok